dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/Gemini/Clients/GeminiChatCompletionClient.cs

```diff
@@ -0,0 +1,581 @@
+            choiceIndex: message.Metadata!.Index,
+            metadata: message.Metadata);
+    }
+
+    private static void ValidateAutoInvoke(bool autoInvoke, int resultsPerPrompt)
+    {
+        if (autoInvoke && resultsPerPrompt != 1)
+        {
+            // We can remove this restriction in the future if valuable. However, multiple results per prompt is rare,
+            // and limiting this significantly curtails the complexity of the implementation.
+            throw new ArgumentException(
+                $"Auto-invocation of tool calls may only be used with a {nameof(GeminiPromptExecutionSettings.CandidateCount)} of 1.");
+        }
+    }
+
+    private static GeminiMetadata GetResponseMetadata(
+        GeminiResponse geminiResponse,
+        GeminiResponseCandidate candidate) => new()
+        {
+            FinishReason = candidate.FinishReason,
+            Index = candidate.Index,
+            PromptTokenCount = geminiResponse.UsageMetadata?.PromptTokenCount ?? 0,
+            CurrentCandidateTokenCount = candidate.TokenCount,
+            CandidatesTokenCount = geminiResponse.UsageMetadata?.CandidatesTokenCount ?? 0,
+            TotalTokenCount = geminiResponse.UsageMetadata?.TotalTokenCount ?? 0,
+            PromptFeedbackBlockReason = geminiResponse.PromptFeedback?.BlockReason,
+            PromptFeedbackSafetyRatings = geminiResponse.PromptFeedback?.SafetyRatings.ToList(),
+            ResponseSafetyRatings = candidate.SafetyRatings?.ToList(),
+        };
+
+    private void LogUsageMetadata(GeminiMetadata metadata)
+    {
+        if (metadata.TotalTokenCount <= 0)
+        {
+            this.Logger.LogDebug("Gemini usage information is not available.");
+            return;
+        }
+
+        this.Logger.LogDebug(
+            "Gemini usage metadata: Candidates tokens: {CandidatesTokens}, Prompt tokens: {PromptTokens}, Total tokens: {TotalTokens}",
+            metadata.CandidatesTokenCount,
+            metadata.PromptTokenCount,
+            metadata.TotalTokenCount);
+
+        s_promptTokensCounter.Add(metadata.PromptTokenCount);
+        s_completionTokensCounter.Add(metadata.CandidatesTokenCount);
+        s_totalTokensCounter.Add(metadata.TotalTokenCount);
+    }
+
+    private sealed class ChatCompletionState
+    {
+        public ChatHistory ChatHistory { get; set; } = null!;
+        public GeminiRequest GeminiRequest { get; set; } = null!;
+        public Kernel Kernel { get; set; } = null!;
+        public GeminiPromptExecutionSettings ExecutionSettings { get; set; } = null!;
+        public GeminiChatMessageContent? LastMessage { get; set; }
+        public int Iteration { get; set; }
+        public bool AutoInvoke { get; set; }
+
+        public void AddLastMessageToChatHistoryAndRequest()
+        {
+            Verify.NotNull(this.LastMessage);
+            this.ChatHistory.Add(this.LastMessage);
+            this.GeminiRequest.AddChatMessage(this.LastMessage);
+        }
```

Comment:
internalize those properties and method.

```suggestion
        internal ChatHistory ChatHistory { get; set; } = null!;
        internal GeminiRequest GeminiRequest { get; set; } = null!;
        internal Kernel Kernel { get; set; } = null!;
        internal GeminiPromptExecutionSettings ExecutionSettings { get; set; } = null!;
        internal GeminiChatMessageContent? LastMessage { get; set; }
        internal int Iteration { get; set; }
        internal bool AutoInvoke { get; set; }

        internal void AddLastMessageToChatHistoryAndRequest()
        {
            Verify.NotNull(this.LastMessage);
            this.ChatHistory.Add(this.LastMessage);
            this.GeminiRequest.AddChatMessage(this.LastMessage);
        }
```

Reply:
Changed but ChatCompletionState was already private.